server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/OrdinalValuesSource.java

```diff
@@ -283,63 +296,86 @@ LeafBucketCollector getLeafCollector(Comparable<BytesRef> value, LeafReaderConte
             throw new IllegalArgumentException("Expected BytesRef, got " + value.getClass());
         }
         BytesRef term = (BytesRef) value;
-        final SortedSetDocValues dvs = docValuesFunc.apply(context);
+        final SortedSetDocValues newLookup = dvsLookup.computeIfAbsent(context.ord, k -> {
+            try {
+                return docValuesFunc.apply(context);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        });
         if (leafReaderContextChanged) {
-            remapOrdinals(lookup, dvs);
-            leafReaderOrd = context.ord;
+            remapOrdinals(lookup, newLookup);
+            lookup = newLookup;
         }
-        lookup = dvs;
+        currentValueOrd = lookup.lookupTerm(term);
+        currentValueUnmapped = null;
+        leafReaderOrd = context.ord;
+        assert currentValueOrd >= 0;
         assert leafReaderContextChanged == false || invariant(); // for performance reasons only check invariant upon change
         return new LeafBucketCollector() {
-            boolean currentValueIsSet = false;
-
             @Override
             public void collect(int doc, long bucket) throws IOException {
                 // caller of getLeafCollector ensures that collection happens before requesting a new leaf collector
                 // this is important as ordinals only make sense in the context of the current lookup
-                assert dvs == lookup;
-                if (currentValueIsSet == false) {
-                    if (dvs.advanceExact(doc)) {
-                        long ord;
-                        while ((ord = dvs.nextOrd()) != NO_MORE_ORDS) {
-                            if (term.equals(dvs.lookupOrd(ord))) {
-                                currentValueIsSet = true;
-                                currentValueOrd = ord;
-                                currentValueUnmapped = null;
-                                break;
-                            }
-                        }
-                    }
-                }
-                assert currentValueIsSet;
                 next.collect(doc, bucket);
             }
         };
     }
 
+    private static class SlotAndOrd implements Comparable<SlotAndOrd> {
+        final int slot;
+        final long ord;
+
+        SlotAndOrd(int slot, long ord) {
+            this.slot = slot;
+            this.ord = ord;
+        }
+
+        @Override
+        public int compareTo(SlotAndOrd other) {
+            long norm1 = ord < 0 ? -ord - 1 : ord;
+            long norm2 = other.ord < 0 ? -other.ord - 1 : other.ord;
+            int cmp = Long.compare(norm1, norm2);
+            return cmp == 0 ? Long.compare(ord, other.ord) : cmp;
```

Comment:
In case we have two negative ords here that are equal, should we fall back to a comparison of unmapped values in order to bring them in order? Otherwise, in case of super high cardinality fields, we might still do random access.

Reply:
Good idea, I pushed https://github.com/elastic/elasticsearch/pull/78313/commits/18ee058c3e2862aa2f99f732e27b53a423ced38b